api/README.md

```diff
@@ -0,0 +1,264 @@
+>
+> Members of CSIProxyService API may be deprecated and then removed from csi-proxy.exe in a manner similar to Kubernetes deprecation [policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/) although maintainers will make an effort to ensure such deprecation is as rare as possible. After their announced deprecation, a member of CSIProxyService API must be supported:
+>
+> 1. 12 months or 3 releases (whichever is longer) if the API member is part of a Stable/vN version.
+> 2. 9 months or 3 releases (whichever is longer) if the API member is part of a Beta/vNbeta1 version.
+> 3. 0 releases if the API member is part of an Alpha/vNalpha1 version.
+> To continue running CSI Node Plugins that depend on an old version of csi-proxy.exe, vN, some of whose members have been removed, Kubernetes administrators will be required to run the latest version of the csi-proxy.exe (that will be used by CSI Node Plugins that use versions of CSIProxyService more recent than vN) along with an old version of csi-proxy.exe that does support vN.
+
+With that in mind, each subsection below details how to deprecate, then remove:
+
+#### A field in an API object
+
+Mark it as deprecated in the `proto` file, e.g.:
+```proto
+message ComputeDoubleRequest{
+    int32 input32 = 1 [deprecated=true];
+}
+```
+then regenerate the protobuf code.
+
+For removal, simply remove it in the first API version that doesn't support that field any more.
+
+#### An API procedure
+
+Similarly, mark the procedure as deprecated in the protobuf definition, e.g.:
+```proto
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {
+        option deprecated = true;
+    }
+}
+```
+then regenerate the protobuf code, and remove it from the first API version that doesn't support that procedure any more.
+
+#### An API version
+
+Again, mark the version as deprecated in its protobuf definition, e.g.:
+```proto
+// Deprecated: Do not use.
+// v1alpha1 is no longer maintained, and will be removed soon.
+package v1alpha1;
+
+service Dummy {
+    ...
+}
+```
+then regenerate the protobuf code, and run `csi-proxy-gen`: it will also mark the version's server (``) and client (``) packages as deprecated.
+
+For removal, remove the whole `api/<api_group_name>/<version>` directory, and run `csi-proxy-gen`, it will remove all references to the removed version.
+
+#### An API group
+
+Deprecate and remove all its versions as explained in the previous version; then remove the entire `api/<api_group_name>` directory, and run `csi-proxy-gen`, it will remove all references to the removed API group.
+
+## Detailed breakdown of generated files
```

Comment:
Can you also add a section on exporting client files? One problem we have with the snapshot APIs + dep, is that users that want the client libraries need to import the external-snapshotter repo, which also contains the server side dependencies, which often include k/k libraries. For the user, dependency management is difficult because now they have dependencies on kubernetes and have to potentially deal with conflicting libraries that may also have transient dependencies on kubernetes.

I'm not sure if go modules can solve this problem, ie exporting the client library as a module. Can we investigate this? Otherwise, we may want to go a similar route with how k/k does it by having a staging directory and then mirroring it to another repo.

Reply:
Good point. In a quick prototype of csi-proxy, I placed the API Go files under a separate package: https://github.com/ddebroy/csi-proxy/tree/master/pkg from the server installation.

The dependency solver was able to import in just the API pkg (v1alpha1) from the repo for when trying it out with gce-pd: https://github.com/ddebroy/gcp-compute-persistent-disk-csi-driver/commit/e129ffdd76f737226aa3bb44af117fee421e54ec#diff-bd247e83efc3c45ae9e8c47233249f18R52

This allowed the vendor repo to be pretty clean on the client repo (gce-pd in this case): https://github.com/ddebroy/gcp-compute-persistent-disk-csi-driver/tree/win-csi1/vendor/github.com/ddebroy/csi-proxy

@wk8 if the above just works with the .proto + generated pb.go files corresponding to the APIs, we can call it out here.

Reply:
Took a quick look at your client api files, it looks like it doesn't depend on any kubernetes imports so maybe it's fine.

Reply:
Digging more into this, it does seem that this might indeed be a problem. Importing any package from a module does import all of that module's dependencies, even if there are not used for the one package you import.

I'm not so worried about k8s imports, as this repo should never import anything from k8s, but it could indeed be a problem anyway: it already has a number of dependencies, there will most likely be many more to come, and ideally we'd like clients to be able to import what they need without getting all the deps with it.

Going to try having 3 separate modules: one under /api, one under /client, and one for the rest; clients would just need to import the 1st two afterwards.

Reply:
Added commit https://github.com/kubernetes-csi/csi-proxy/pull/7/commits/5d2a22bd5373c6c7afbb893ba3d81b3932f856ee in which everything that clients need to import is broken into its own go module.